### ISSUE TITLE
When libp2p fails, node never re-attempts to get connection

### ISSUE THREAD

**Issue by SjonHortensius** [prysmaticlabs/prysm#2674]
Not sure what caused this - but it might be a temporary network glitch:

```
[2019-05-22 10:57:10]  INFO initial-sync: Synced!
[2019-05-22 10:57:10]  INFO regular-sync: Listening for regular sync messages from peers
[2019-05-22 15:30:39] ERROR p2p: Failed to reconnect to peer failed to dial : all dials failed
  * [/ip4/23.202.xxx/tcp/4000] dial tcp4 23.202.xxx:4000: connect: connection refused
  * [/ip4/23.195.xxx/tcp/4000] dial tcp4 23.195.xxx:4000: connect: connection refused
  * [/ip4/23.217.xxx/tcp/4000] dial tcp4 23.217.xxx:4000: connect: connection refused
  * [/ip4/127.0.0.1/tcp/4000] failed to negotiate security protocol: message did not have trailing newline
  * [/ip4/23.217.xxx/tcp/4000] dial tcp4 23.217.xxx:4000: connect: connection refused
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 10.52.xxx:4000: connect: no route to host
  * [/ip4/35.224.xxx/tcp/30001] failed to negotiate security protocol: EOF
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 10.52.xxx:4000: connect: no route to host
  * [/ip4/10.55.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->10.55.xxx:4000: i/o timeout
  * [/ip4/92.242.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->92.242.xxx:4000: i/o timeout
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->10.52.xxx:4000: i/o timeout
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->10.52.xxx:4000: i/o timeout
  * [/ip4/18.211.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->18.211.xxx:4000: i/o timeout
[2019-05-22 15:30:52] ERROR p2p: Failed to reconnect to peer failed to dial : all dials failed
  * [/ip4/23.202.xxx/tcp/4000] dial tcp4 23.202.xxx:4000: connect: connection refused
  * [/ip4/23.195.xxx/tcp/4000] dial tcp4 23.195.xxx:4000: connect: connection refused
  * [/ip4/127.0.0.1/tcp/4000] failed to negotiate security protocol: message did not have trailing newline
  * [/ip4/23.217.xxx/tcp/4000] dial tcp4 23.217.xxx:4000: connect: connection refused
  * [/ip4/23.217.xxx/tcp/4000] dial tcp4 23.217.xxx:4000: connect: connection refused
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 10.52.2.193:4000: connect: no route to host
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 10.52.4.162:4000: connect: no route to host
  * [/ip4/92.242.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->92.242.xxx:4000: i/o timeout
  * [/ip4/10.55.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->10.55.xxx:4000: i/o timeout
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->10.52.xxx:4000: i/o timeout
  * [/ip4/10.52.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->10.52.xxx:4000: i/o timeout
  * [/ip4/18.211.xxx/tcp/4000] dial tcp4 0.0.0.0:12000->18.211.xxx:4000: i/o timeout
  * [/ip4/35.224.xxx/tcp/30001] failed to negotiate security protocol: read tcp4 192.168.xxx:12000->35.224.xxx/30001: read: connection reset by peer
[2019-05-22 15:30:58] ERROR p2p: Failed to reconnect to peer dial backoff
```

I'm see a few things wrong with this:
* after this error the node hangs and never attempts to reconnect. It should schedule a reconnect after a few minutes instead.
* this might also indicate an issue in #2245 as 10.x is not in my list of reachable networks - these nodes should be skipped
* `127.0.0.1` should not be in this list

**Comment by prestonvanloon** [comment]
Blocked by
- https://github.com/libp2p/go-libp2p-kad-dht/issues/330
- https://github.com/libp2p/go-libp2p-kad-dht/issues/302

**Comment by dimchansky** [comment]
Not sure about this case, but this pattern helped me to solve re-connection issue in my app:

```go
host.Peerstore().ClearAddrs(targetPeerID)

if sw, ok := host.Network().(*swarm.Swarm); ok {
	sw.Backoff().Clear(targetPeerID)
}

if err := host.Connect(ctx, *targetPeerAddr); err != nil {
	return nil, err
}
```

**Comment by rauljordan** [comment]
No longer relevant in latest master
